tasks/fullStack/app/controllers/reviews_controller.rb

```diff
@@ -3,27 +3,45 @@ class ReviewsController < ApplicationController
   DEFAULT_TAGS = ['default']
 
   def index
-    if params[:shop_id].present? && Shop.where("id = #{params[:shop_id]}").present?
-      params[:per_page] ||= 10
-      offset = params[:page].to_i * params[:per_page]
-
-      @data = []
-      products = Product.where("shop_id = #{params[:shop_id]}").sort_by(&:created_at)[offset..(offset + params[:per_page])]
-      products.each do |product|
-        reviews = product.reviews.sort_by(&:created_at)[offset..(offset + params[:per_page])]
-        @data << { product: product, reviews: reviews }
-      end
+    params[:per_page] ||= 10
+    offset = params[:page].to_i * params[:per_page]
+
+    products_query = Product.includes(:reviews).order(created_at: :desc)
+    products_query = products_query.where(shop_id: params[:shop_id].to_i) if params[:shop_id].present?
+
+    products = products_query.limit(params[:per_page].to_i).offset(offset)
+    product_ids = products.pluck(:id)
+
+    reviews = Review.where(product_id: product_ids).group_by(&:product_id)
+
+    @data = products.map do |product|
+      {
+        product: product,
+        reviews: reviews[product.id].try(:sort_by, &:created_at).try(:first, 3) || []
+      }
     end
   end
 
   def create
     # TODO: Create reviews in background. No need to show errors (if any) to users, it's fine to skip creating the review silently when some validations fail.
 
     tags = tags_with_default(params)
-    Review.create(product_id: params[:product_id], body: params[:body], rating: params[:rating], reviewer_name: params[:reviewer_name], tags: tags)
+
+    CreateReviewWorker.perform_async({
+      "product_id" => params[:review][:product_id].to_i,
+      "body" => params[:review][:body],
+      "rating" => params[:review][:rating].to_i,
+      "reviewer_name" => params[:review][:reviewer_name],
```

Comment:
Accessing review parameters would be easier and safer with permitted params

```ruby
private
...
def review_params
  params.require(:review).permit(:product_id, :body, :rating, :reviewer_name)
end
```

```suggestion
    CreateReviewWorker.perform_async({
      "product_id" => review_params[:product_id].to_i,
      "body" => review_params[:body],
      "rating" => review_params[:rating].to_i,
      "reviewer_name" => review_params[:reviewer_name],
```

Reply:
ah I see! I hadn't been exposed to the concept of strong parameters before but this makes sense both for the sake of security and readability.